File: Sources/Services/FixtureMarketDataService.swift
import Foundation

protocol LongbridgeDataServicing {
    func loadFixtureSnapshot() throws -> LongbridgeDataSnapshot
}

enum FixtureDataError: Error {
    case missingResource(String)
}

final class FixtureLongbridgeDataService: LongbridgeDataServicing {
    private let cache: MarketDataCache
    private let universeService: UniverseServicing
    private let bundle: Bundle
    private let decoder: JSONDecoder

    init(
        cache: MarketDataCache,
        universeService: UniverseServicing,
        bundle: Bundle = .main
    ) {
        self.cache = cache
        self.universeService = universeService
        self.bundle = bundle

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        self.decoder = decoder
    }

    func loadFixtureSnapshot() throws -> LongbridgeDataSnapshot {
        let capabilities: LongbridgeCapabilities = try read("cli-capabilities")
        let authorization: AuthorizationStatusSummary = try read(
            "authorization-status"
        )
        let connectivity: ConnectivitySummary = try read("connectivity-check")
        let historicalBars: HistoricalBarSeries = try read(
            "historical-daily-bars"
        )
        let currentSnapshot: CurrentMarketSnapshot = try read(
            "current-market-snapshot"
        )
        let marketStatus: MarketStatusSnapshot = try read("market-status")
        let securityList: SecurityListSnapshot = try read("security-list")
        let positions: BrokerPositionSnapshot = try read("broker-positions")
        let configuration: UniverseConfiguration = try read("universe-config")
        let universe = universeService.buildDailySnapshot(
            snapshotTime: securityList.collectedAt,
            configuration: configuration,
            securities: securityList,
            positions: positions
        )

        _ = try cache.storeHistoricalBars(historicalBars)
        _ = try cache.storeCurrentSnapshot(currentSnapshot)
        _ = try cache.storeUniverseSnapshot(universe)

        return LongbridgeDataSnapshot(
            capabilities: capabilities,
            authorization: authorization,
            connectivity: connectivity,
            historicalBars: historicalBars,
            currentSnapshot: currentSnapshot,
            marketStatus: marketStatus,
            securityList: securityList,
            brokerPositions: positions,
            universe: universe
        )
    }

    private func read<T: Decodable>(_ name: String) throws -> T {
        let url = bundle.url(
            forResource: name,
            withExtension: "json",
            subdirectory: "fixtures/longbridge"
        ) ?? bundle.resourceURL?
            .appendingPathComponent("fixtures/longbridge", isDirectory: true)
            .appendingPathComponent("\(name).json")
        guard let url, FileManager.default.fileExists(atPath: url.path) else {
            throw FixtureDataError.missingResource(name)
        }
        return try decoder.decode(T.self, from: Data(contentsOf: url))
    }
}

File: Windows/Services/FixtureMarketDataService.cs
using System.IO;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CytisusTrading.Windows;

public interface ILongbridgeDataService
{
    LongbridgeDataSnapshot LoadFixtureSnapshot();
}

public sealed class FixtureLongbridgeDataService : ILongbridgeDataService
{
    private readonly IMarketDataCache _cache;
    private readonly IUniverseService _universeService;
    private readonly Assembly _assembly;
    private readonly JsonSerializerOptions _options;

    public FixtureLongbridgeDataService(
        IMarketDataCache cache,
        IUniverseService universeService,
        Assembly? assembly = null)
    {
        _cache = cache;
        _universeService = universeService;
        _assembly = assembly ?? typeof(FixtureLongbridgeDataService).Assembly;
        _options = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            PropertyNameCaseInsensitive = true
        };
        _options.Converters.Add(new JsonStringEnumConverter());
    }

    public LongbridgeDataSnapshot LoadFixtureSnapshot()
    {
        var capabilities = Read<LongbridgeCapabilities>("cli-capabilities.json");
        var authorization = Read<AuthorizationStatusSummary>("authorization-status.json");
        var connectivity = Read<ConnectivitySummary>("connectivity-check.json");
        var bars = Read<HistoricalBarSeries>("historical-daily-bars.json");
        var current = Read<CurrentMarketSnapshot>("current-market-snapshot.json");
        var marketStatus = Read<MarketStatusSnapshot>("market-status.json");
        var securities = Read<SecurityListSnapshot>("security-list.json");
        var positions = Read<BrokerPositionSnapshot>("broker-positions.json");
        var configuration = Read<UniverseConfiguration>("universe-config.json");
        var universe = _universeService.BuildDailySnapshot(
            securities.CollectedAt,
            configuration,
            securities,
            positions);

        _cache.StoreHistoricalBars(bars);
        _cache.StoreCurrentSnapshot(current);
        _cache.StoreUniverseSnapshot(universe);

        return new LongbridgeDataSnapshot(
            capabilities,
            authorization,
            connectivity,
            bars,
            current,
            marketStatus,
            securities,
            positions,
            universe);
    }

    private T Read<T>(string fileName)
    {
        var resourceName = _assembly
            .GetManifestResourceNames()
            .SingleOrDefault(name =>
                name.EndsWith(fileName, StringComparison.OrdinalIgnoreCase));
        if (resourceName is null)
        {
            throw new FileNotFoundException(
                $"Embedded Longbridge fixture not found: {fileName}");
        }

        using var stream = _assembly.GetManifestResourceStream(resourceName)
            ?? throw new FileNotFoundException(
                $"Embedded Longbridge fixture stream not found: {fileName}");
        return JsonSerializer.Deserialize<T>(stream, _options)
            ?? throw new InvalidDataException(
                $"Invalid Longbridge fixture JSON: {fileName}");
    }
}
